### DISCUSSION TITLE
Building the UNL using strong p2p overlays such as Kelips can lead to security and speed benefits

### DISCUSSION THREAD

**Discussion by armor009** [ripple/rippled#4273] [Ideas]
Building the UNL using strong p2p overlays such as Kelips can lead to security benefits. The network receives strong and fast information propagation properties compared to a network with random UNLs. Vulnerabilities like information and transaction censorship,  network partitions, blockchain forks  can be reduced significantly. Squelch ammendment can be applied to reduce the messages sent. I'm working on a research paper using this.

This is an old version of the paper

https://arxiv.org/abs/2008.01742

**Comment by MarkusTeufelberger** [comment]
UNLs don't have anything to do with the p2p layer... I'm not sure you fully understand the terminology used in the XRPL space.

**Reply by armor009** [reply]
Yes. I'm aware about the terminologies in the XRPL Space, particularly those associated with the XRP LCP. :)

As a step 1. Was thinking of applying Kelips as a p2p layer to the gossip layer. The next step would be application to the UNL.

As a next step I am suggesting that we apply to the UNL - Kelips' structure. This will bring in censorship resilience, reduce blockchain fork chances, auto UNL design, churn resilience, protection from eclipse attack, protection from table poisoning. It would make it very difficult for several  malicious nodes to gain combined entry into the UNL. It will boost speed, security, last mile connectivity, and ease of use.

This is based on the original definition of UNL - a list of servers which a server participating in XRP Ledger Consensus believes will not collectively defraud it and which represents the entire network.

We can even keep the starter UNL as one of the components of the UNL

Currently there is no mechanism/ guideline to ensure overlap amongst nodes apart from the starter UNL. If one is able to find out the make up of a node's UNL, it becomes easier to attack the node. However by bringing in mechanisms and probabilistic protection, it becomes difficult to attack the nodes while still having of safety and  the notion 'UNL overlap'.

Obviously for churn resilience, the max 10% changes in UNL rule can be applied

**Comment by cjcobb23** [comment]
What exactly do you mean by "building" the UNL? The UNL is published by a UNL publisher, not built. Also, what exactly do you mean by "random UNLs"? UNLs are not random; each node operator selects the UNL they want to use, and the UNL publisher selects validators to put on their UNL.

**Reply by armor009** [reply]
The UNL is a necessary part for the network to achieve consensus. It is based on the concept of collective trust. It is assumed that the members of the UNL will not collectively defraud the network. The UNL represents the network as far as a server is considered. The UNL is maintained or chosen by the server's administrator. There is a starter UNL published by Ripple. Most people use the UNL published by Ripple. In the future, as the network matures, it is expected that other people share their proposed starter UNL for all to see. 

There is also the concept of peer discovery which helps add more members to the server's UNL.

Technically it is a security vulnerability if a server discloses its UNL. That can open the server to eclipse attacks or censorship. As servers can then attempt to game the system.

The servers in the current starter UNL have a high connection degree. They serve as a network backbone.

For achieving safety properties, the UNLs of all nodes need 90-99% overlap. This is highly restrictive. There currently is no mechanism (other than the starter UNL) to ensure that this overlap is met.

Sharing UNLs is a security vulnerability as previously stated. This makes it difficult to ensure necessary overlaps.

In the research paper I have mentioned an approach/guideline to setting up the UNL. It relies on the original definition of UNL (stated above).

It makes use of the concept of p2p network overlays and message passing to satisfy the UNL's definition.

The comparison with random UNL is used in ripple's simulator that shows liveness. 

The p2p approach that I propose also involves randomisation and probabilistic protection against eclipse and censorship. It makes it very difficult for a large number of malicious nodes to enter a node's UNL. The only case where that would happen is if there are more malicious servers than genuine servers. 10% change in UNL rule resolves it.

The p2p overlay concept can also be used in conjunction with the starter UNL. Putting the starter UNL into one of the affinity groups or spreading it across affinity groups.

The research paper shared is an old version. I am currently splitting  the research paper into 2 papers. First one improves the p2p gossip layer. Second one is envisioned to work on the UNL selection.

The Chase Macbrough paper on XRP LCP also identities the need of an approach that makes use of an understanding of the network's message passing behaviour.

**Comment by armor009** [comment]
Would love to collaborate in improving the Ripple system and publishing a research paper
